In [3]:
import pandas as pd
import pyarrow.parquet as pq
from time import time

In [5]:
pq.read_metadata('yellow_tripdata_2021-01.parquet')

  created_by: parquet-cpp-arrow version 7.0.0
  num_columns: 19
  num_rows: 1369769
  num_row_groups: 1
  format_version: 1.0
  serialized_size: 10382

In [6]:
# Read file, read the table from file and check schema
file = pq.ParquetFile('yellow_tripdata_2021-01.parquet')
table = file.read()
table.schema

VendorID: int64
tpep_pickup_datetime: timestamp[us]
tpep_dropoff_datetime: timestamp[us]
passenger_count: double
trip_distance: double
RatecodeID: double
store_and_fwd_flag: string
PULocationID: int64
DOLocationID: int64
payment_type: int64
fare_amount: double
extra: double
mta_tax: double
tip_amount: double
tolls_amount: double
improvement_surcharge: double
total_amount: double
congestion_surcharge: double
airport_fee: double
-- schema metadata --
pandas: '{"index_columns": [], "column_indexes": [], "columns": [{"name":' + 2492

In [7]:
# Convert to pandas and check data 
df = table.to_pandas()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1369769 entries, 0 to 1369768
Data columns (total 19 columns):
 #   Column                 Non-Null Count    Dtype         
---  ------                 --------------    -----         
 0   VendorID               1369769 non-null  int64         
 1   tpep_pickup_datetime   1369769 non-null  datetime64[us]
 2   tpep_dropoff_datetime  1369769 non-null  datetime64[us]
 3   passenger_count        1271417 non-null  float64       
 4   trip_distance          1369769 non-null  float64       
 5   RatecodeID             1271417 non-null  float64       
 6   store_and_fwd_flag     1271417 non-null  object        
 7   PULocationID           1369769 non-null  int64         
 8   DOLocationID           1369769 non-null  int64         
 9   payment_type           1369769 non-null  int64         
 10  fare_amount            1369769 non-null  float64       
 11  extra                  1369769 non-null  float64       
 12  mta_tax                13697

We need to first create the connection to our postgres database. We can feed the connection information to generate the CREATE SQL query for the specific server. SQLAlchemy supports a variety of servers.

In [10]:
# Create an open SQL database connection object or a SQLAlchemy connectable
from sqlalchemy import create_engine

engine = create_engine('postgresql://root:root@localhost:5432/ny_taxi')
engine.connect()

In [11]:
# Generate CREATE SQL statement from schema for validation
print(pd.io.sql.get_schema(df, name='yellow_taxi_data', con=engine))


CREATE TABLE yellow_taxi_data (
	"VendorID" BIGINT, 
	tpep_pickup_datetime TIMESTAMP WITHOUT TIME ZONE, 
	tpep_dropoff_datetime TIMESTAMP WITHOUT TIME ZONE, 
	passenger_count FLOAT(53), 
	trip_distance FLOAT(53), 
	"RatecodeID" FLOAT(53), 
	store_and_fwd_flag TEXT, 
	"PULocationID" BIGINT, 
	"DOLocationID" BIGINT, 
	payment_type BIGINT, 
	fare_amount FLOAT(53), 
	extra FLOAT(53), 
	mta_tax FLOAT(53), 
	tip_amount FLOAT(53), 
	tolls_amount FLOAT(53), 
	improvement_surcharge FLOAT(53), 
	total_amount FLOAT(53), 
	congestion_surcharge FLOAT(53), 
	airport_fee FLOAT(53)
)




## Finally inserting data

In [12]:
#This part is for testing


# Creating batches of 100,000 for the paraquet file
batches_iter = file.iter_batches(batch_size=100000)
batches_iter

# Take the first batch for testing
df = next(batches_iter).to_pandas()
df

# Creating just the table in postgres
#df.head(0).to_sql(name='ny_taxi_data',con=engine, if_exists='replace')

VendorID tpep_pickup_datetime tpep_dropoff_datetime  passenger_count  \
0             1  2021-01-01 00:30:10   2021-01-01 00:36:12              1.0   
1             1  2021-01-01 00:51:20   2021-01-01 00:52:19              1.0   
2             1  2021-01-01 00:43:30   2021-01-01 01:11:06              1.0   
3             1  2021-01-01 00:15:48   2021-01-01 00:31:01              0.0   
4             2  2021-01-01 00:31:49   2021-01-01 00:48:21              1.0   
...         ...                  ...                   ...              ...   
99995         1  2021-01-04 14:04:31   2021-01-04 14:08:52              3.0   
99996         1  2021-01-04 14:18:46   2021-01-04 14:35:45              2.0   
99997         1  2021-01-04 14:42:41   2021-01-04 14:59:22              2.0   
99998         2  2021-01-04 14:39:02   2021-01-04 15:09:37              2.0   
99999         2  2021-01-04 14:49:36   2021-01-04 14:54:44              5.0   

       trip_distance  RatecodeID store_and_fwd_flag  PULocationID  \
0               2.10         1.0                  N           142   
1               0.20         1.0                  N           238   
2              14.70         1.0                  N           132   
3              10.60         1.0                  N           138   
4               4.94         1.0                  N            68   
...              ...         ...                ...           ...   
99995           0.70         1.0                  N           234   
99996           3.30         1.0                  N           234   
99997           4.70         1.0                  N           236   
99998          17.95         2.0                  N           132   
99999           0.37         1.0                  N           236   

       DOLocationID  payment_type  fare_amount  extra  mta_tax  tip_amount  \
0                43             2          8.0    3.0      0.5        0.00   
1               151             2          3.0    0.5      0.5        0.00   
2               165             1         42.0    0.5      0.5        8.65   
3               132             1         29.0    0.5      0.5        6.05   
4                33             1         16.5    0.5      0.5        4.06   
...             ...           ...          ...    ...      ...         ...   
99995           224             2          5.0    2.5      0.5        0.00   
99996           236             1         14.5    2.5      0.5        3.55   
99997            79             1         17.0    2.5      0.5        4.05   
99998           148             1         52.0    0.0      0.5        5.00   
99999           236             2          5.0    0.0      0.5        0.00   

       tolls_amount  improvement_surcharge  total_amount  \
0               0.0                    0.3         11.80   
1               0.0                    0.3          4.30   
2               0.0                    0.3         51.95   
3               0.0                    0.3         36.35   
4               0.0                    0.3         24.36   
...             ...                    ...           ...   
99995           0.0                    0.3          8.30   
99996           0.0                    0.3         21.35   
99997           0.0                    0.3         24.35   
99998           0.0                    0.3         60.30   
99999           0.0                    0.3          8.30   

       congestion_surcharge  airport_fee  
0                       2.5          NaN  
1                       0.0          NaN  
2                       0.0          NaN  
3                       0.0          NaN  
4                       2.5          NaN  
...                     ...          ...  
99995                   2.5          NaN  
99996                   2.5          NaN  
99997                   2.5          NaN  
99998                   2.5          NaN  
99999                   2.5          NaN  

[100000 rows x 19 columns]

There are 1369769 rows in our dataset. We are going to use the parquet_file.iter_batches() function to create batches of 100,000, convert them into pandas and then load it into the postgres database.

In [13]:
t_start = time()
count = 0
for batch in file.iter_batches(batch_size=100000):
    count+=1
    batch_df = batch.to_pandas()
    print(f'inserting batch {count}...')
    b_start = time()
    
    batch_df.to_sql(name='yellow_taxi_data',con=engine, if_exists='append')
    b_end = time()
    print(f'inserted! time taken {b_end-b_start:10.3f} seconds.\n')
    
t_end = time()   
print(f'Completed! Total time taken was {t_end-t_start:10.3f} seconds for {count} batches.')

inserting batch 1...
inserted! time taken     19.204 seconds.

inserting batch 2...
inserted! time taken     24.590 seconds.

inserting batch 3...
inserted! time taken     16.425 seconds.

inserting batch 4...
inserted! time taken     24.599 seconds.

inserting batch 5...
inserted! time taken     24.206 seconds.

inserting batch 6...
inserted! time taken     29.109 seconds.

inserting batch 7...
inserted! time taken     22.158 seconds.

inserting batch 8...
inserted! time taken     16.741 seconds.

inserting batch 9...
inserted! time taken     18.121 seconds.

inserting batch 10...
inserted! time taken     24.329 seconds.

inserting batch 11...
inserted! time taken     23.103 seconds.

inserting batch 12...
inserted! time taken     23.222 seconds.

inserting batch 13...
inserted! time taken     18.506 seconds.

inserting batch 14...
inserted! time taken     10.982 seconds.

Completed! Total time taken was    295.656 seconds for 14 batches.
